Pegasus/sources/Scene.cpp

```diff
@@ -4,109 +4,86 @@
-#include <pegasus/Collision.hpp>
+#include <pegasus/Debug.hpp>
+#include <pegasus/Force.hpp>
 #include <pegasus/Integration.hpp>
 
 namespace pegasus
 {
 namespace scene
 {
 
-RigidBody::RigidBody(Handle body, Handle shape)
-    : body(body)
-    , shape(shape)
+Scene::Scene()
+    : m_detector(m_assetManager)
+    , m_resolver(m_assetManager)
 {
 }
 
-AssetManager& AssetManager::GetInstance()
+void Scene::ComputeFrame(double duration)
 {
-    static AssetManager am;
-    return am;
-}
+    ApplyCollisionCache(duration);
 
-std::vector<Asset<mechanics::Body>>& AssetManager::GetBodies()
-{
-    return m_bodies;
-}
+    ApplyForces(1);
 
-StaticBody::StaticBody(Handle body, Handle shape)
-    : RigidBody(body, shape)
-{
-    static AssetManager& am = AssetManager::GetInstance();
-    am.GetAsset(am.GetBodies(), body).material.SetInfiniteMass();
-}
+    Integrate(duration);
 
-DynamicBody::DynamicBody(Handle body, Handle shape)
-    : RigidBody(body, shape)
-{
+    ResolveCollisions(duration);
 }
 
-Scene& Scene::GetInstance()
+Handle Scene::MakeBody()
 {
-    static Scene scene;
-    return scene;
+    return m_assetManager.MakeAsset(m_assetManager.GetBodies());
 }
 
-void Scene::Initialize(AssetManager& assetManager)
+mechanics::Body& Scene::GetBody(Handle handle)
 {
-    m_assetManager = &assetManager;
+    return m_assetManager.GetAsset(m_assetManager.GetBodies(), handle);
 }
 
-void Scene::ComputeFrame(double duration)
+void Scene::RemoveBody(Handle handle)
 {
-    ResolveCollisions(duration);
-
-    ApplyForces();
-
-    Integrate(duration);
+    m_assetManager.RemoveAsset(m_assetManager.GetBodies(), handle);
 }
 
-Handle Scene::MakeBody() const
+AssetManager& Scene::GetAssets()
 {
-    return m_assetManager->MakeAsset(m_assetManager->GetBodies());
+    return m_assetManager;
 }
 
-mechanics::Body& Scene::GetBody(Handle handle) const
+void Scene::ResolveCollisions(double duration)
 {
-    return m_assetManager->GetAsset(m_assetManager->GetBodies(), handle);
+    std::vector<collision::Contact> contacts = m_detector.Detect();
+    Debug::CollisionDetectionCall(contacts);
+    m_resolver.Resolve(contacts, duration);
 }
 
-void Scene::RemoveBody(Handle handle) const
+void Scene::ApplyCollisionCache(double duration)
 {
-    m_assetManager->RemoveAsset(m_assetManager->GetBodies(), handle);
+    m_resolver.ResolvePersistantContacts(duration);
 }
 
-void Scene::ResolveCollisions(double duration) const
-{
-    static collision::Detector detector(*m_assetManager);
-    std::vector<std::vector<collision::Contact>> contacts = detector.Detect();
-
-    static collision::Resolver resolver;
-    resolver.Resolve(contacts, duration);
-}
-
-void Scene::ApplyForces()
+void Scene::ApplyForces(double duration)
 {
     //Clear previously applied forces
-    for (Asset<mechanics::Body>& asset : m_assetManager->GetBodies())
+    for (Asset<mechanics::Body>& asset : m_assetManager.GetBodies())
     {
         asset.data.linearMotion.force = glm::dvec3(0);
```

Comment:
Shall `angularMotion.torque` be reset as well?

Reply:
yeah

Reply:
good idea